### PR TITLE
feat: JumpStart proprietary manifest/specs parsing

### PR DESCRIPTION
**Pull request by Captainia**
*Issue #, if available:*

*Description of changes:* Add parsing and caching utilities to recognize a new model type: JumpStart proprietary models. The proprietary models are read from a complete separate pair of metadata: proprietary-sdk-manifest.json and each model follows the mapping to a spec file under proprietary-models/<model-id>/proprietary_specs_<model_id>.json.

*Testing done:* unit and integ tests

## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

#### General

- [x] I have read the [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md) doc
- [x] I certify that the changes I am introducing will be backward compatible, and I have discussed concerns about this, if any, with the Python SDK team
- [x] I used the commit message format described in [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md#committing-your-change)
- [x] I have passed the region in to all S3 and STS clients that I've initialized as part of this change.
- [ ] I have updated any necessary documentation, including [READMEs](https://github.com/aws/sagemaker-python-sdk/blob/master/README.rst) and [API docs](https://github.com/aws/sagemaker-python-sdk/tree/master/doc) (if appropriate)

#### Tests

- [ ] I have added tests that prove my fix is effective or that my feature works (if appropriate)
- [ ] I have added unit and/or integration tests as appropriate to ensure backward compatibility of the changes
- [ ] I have checked that my tests are not configured for a specific region or account (if appropriate)
- [ ] I have used [`unique_name_from_base`](https://github.com/aws/sagemaker-python-sdk/blob/master/src/sagemaker/utils.py#L77) to create resource names in integ tests (if appropriate)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by mufaddal-rohawala** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: sagemaker-python-sdk-unit-tests
* Commit ID: eed4edc55b02f67b25d44ab418c077435ab6c8e6
* Result: FAILED
* [Build Logs](https://fegcnxu6da.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=471d9f50-0486-4c59-b9c2-041a47c62c64%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by mufaddal-rohawala** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: sagemaker-python-sdk-notebook-tests
* Commit ID: eed4edc55b02f67b25d44ab418c077435ab6c8e6
* Result: FAILED
* [Build Logs](https://bkqkixfb59.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=532a2fcd-dd25-4b4f-ab44-c90ed798ade7%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by mufaddal-rohawala** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: sagemaker-python-sdk-local-mode-tests
* Commit ID: eed4edc55b02f67b25d44ab418c077435ab6c8e6
* Result: SUCCEEDED
* [Build Logs](https://8huhkqo601.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=68926e34-3edb-4fa0-ae1a-5e6ee5f415a7%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by mufaddal-rohawala** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: sagemaker-python-sdk-unit-tests
* Commit ID: 19a6e3ac78a8d27540067ada3ad51e1892c5c28d
* Result: FAILED
* [Build Logs](https://fegcnxu6da.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=3b07a286-5ae2-481f-8d5d-27a64e75c3d3%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by mufaddal-rohawala** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: sagemaker-python-sdk-unit-tests
* Commit ID: 19a6e3ac78a8d27540067ada3ad51e1892c5c28d
* Result: FAILED
* [Build Logs](https://fegcnxu6da.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=9cb7f0e2-8df8-4aeb-b851-dc1ed95524a5%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by mufaddal-rohawala** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: sagemaker-python-sdk-unit-tests
* Commit ID: 19a6e3ac78a8d27540067ada3ad51e1892c5c28d
* Result: FAILED
* [Build Logs](https://fegcnxu6da.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=4d7e1fb2-eb4c-4e0a-8af9-82584c8fe7ee%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by mufaddal-rohawala** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: sagemaker-python-sdk-unit-tests
* Commit ID: 5ee2907230bce9bab35d40ab7bf94e781e7b2743
* Result: FAILED
* [Build Logs](https://fegcnxu6da.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=e39d8578-f0f7-4281-961e-63df0464e375%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

File: src/sagemaker/jumpstart/cache.py [hunk 1]
```diff
@@ -149,10 +195,10 @@ def get_bucket(self) -> str:
         """Return bucket used for cache."""
         return self.s3_bucket_name
 
-    def _get_manifest_key_from_model_id_semantic_version(
-        self,
+    def _model_id_retrival_function(self,
```
Comment (evakravi) [inline src/sagemaker/jumpstart/cache.py:198]: typo: `retrival`

File: src/sagemaker/jumpstart/cache.py [hunk 2]
```diff
@@ -329,36 +390,42 @@ def _retrieval_function(
                 formatted_content=utils.get_formatted_manifest(formatted_body),
                 md5_hash=etag,
             )
-        if file_type == JumpStartS3FileType.SPECS:
+        if file_type in {JumpStartS3FileType.SPECS, JumpStartS3FileType.PROPRIETARY_SPECS}:
```
Comment (evakravi) [inline src/sagemaker/jumpstart/cache.py:393]: consider creating new constants/enums for just manifest/specs

**Comment by mufaddal-rohawala** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: sagemaker-python-sdk-local-mode-tests
* Commit ID: 5ee2907230bce9bab35d40ab7bf94e781e7b2743
* Result: SUCCEEDED
* [Build Logs](https://8huhkqo601.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=3c530528-3f43-42c8-854f-c54aecd3e07d%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by mufaddal-rohawala** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: sagemaker-python-sdk-notebook-tests
* Commit ID: 5ee2907230bce9bab35d40ab7bf94e781e7b2743
* Result: SUCCEEDED
* [Build Logs](https://bkqkixfb59.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=166e73af-5024-4732-bf6d-8ea4a7147521%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

File: src/sagemaker/jumpstart/cache.py [hunk 3]
```diff
@@ -129,15 +146,44 @@ def get_region(self) -> str:
         """Return region for cache."""
         return self._region
 
-    def set_manifest_file_s3_key(self, key: str) -> None:
-        """Set manifest file s3 key. Clears cache after new key is set."""
-        if key != self._manifest_file_s3_key:
-            self._manifest_file_s3_key = key
-            self.clear()
+    def set_manifest_file_s3_key(
```
Comment (akozd) [inline src/sagemaker/jumpstart/cache.py:149]: this is cleaner IMO
```python
def set_manifest_file_s3_key(
    self,
    key: str,
    file_type: JumpStartS3FileType = JumpStartS3FileType.MANIFEST,
) -> None:
    # mapping is {JumpStartS3FileType.MANIFEST: '_manifest_file_s3_key', etc...}
    if not (property_name := mapping.get(file_type)):
        raise ValueError(
            f"Bad value when setting manifest '{file_type}':" \
            f"must be in {[JumpStartS3FileType.MANIFEST, JumpStartS3FileType.PROPRIETARY_MANIFEST]}"
        )
    if key != getattr(self, property_name, None):
        setattr(self, property_name, key)
        self.clear()
```

File: src/sagemaker/jumpstart/cache.py [hunk 4]
```diff
@@ -129,15 +146,44 @@ def get_region(self) -> str:
...
+        self,
+        key: str,
+        file_type: JumpStartS3FileType = JumpStartS3FileType.MANIFEST,
+    ) -> None:
+        """Set manifest file s3 key. Clears cache after new key is set.
 
-    def get_manifest_file_s3_key(self) -> str:
+        Raises:
+            ValueError: if hte file type is not recognized
+        """
```
Comment (akozd) [inline src/sagemaker/jumpstart/cache.py:158]: along the lines of the comment on set_manifest_file_s3_key, this logic can be simplified

File: src/sagemaker/jumpstart/cache.py [hunk 5]
```diff
@@ -402,14 +470,21 @@ def _get_header_impl(
                 header.
             attempt (int): attempt number at retrieving a header.
         """
+        if model_type == JumpStartModelType.OPENSOURCE:
+            versioned_model_id = self._open_source_model_id_manifest_key_cache.get(
+                JumpStartVersionedModelId(model_id, semantic_version_str)
+            )[0]
```
Comment (akozd) [inline src/sagemaker/jumpstart/cache.py:476]: why the `[0]`?

File: src/sagemaker/jumpstart/utils.py [hunk 6]
```diff
@@ -732,18 +738,33 @@ def resolve_estimator_sagemaker_config_field(
...
-    """Returns True if the model ID is supported for the given script.
+) -> enums.JumpStartModelType:
+    """Returns model type if the model ID is supported for the given script.
 
     Raises:
         ValueError: If the script is not supported by JumpStart.
     """
+    def _get_model_type(
+        model_id: str,
+        open_source_models: List[str],
```
Comment (akozd) [inline src/sagemaker/jumpstart/utils.py:755]: It would be nice for these to be sets for easier O(1) lookup rather than O(n)

**Review by akozd — APPROVED** [review]

**Comment by mufaddal-rohawala** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: sagemaker-python-sdk-pr
* Commit ID: 5ee2907230bce9bab35d40ab7bf94e781e7b2743
* Result: FAILED
* [Build Logs](https://abkt2mlh3k.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=51bc982a-d94a-428f-9524-bdf601cd4bba%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by mufaddal-rohawala** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: sagemaker-python-sdk-slow-tests
* Commit ID: 5ee2907230bce9bab35d40ab7bf94e781e7b2743
* Result: FAILED
* [Build Logs](https://omfrydptb9.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=18f82be1-dbf5-45bf-b0a8-3da8899d2d49%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

File: src/sagemaker/jumpstart/utils.py (hunk 6 above)
Comment (Captainia) [reply src/sagemaker/jumpstart/utils.py]: Yes this is misleading typing, they are sets already. Updated

**Comment by mufaddal-rohawala** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: sagemaker-python-sdk-unit-tests
* Commit ID: ab8ee02a103cc8102540fabe8d3b5471e13f09e4
* Result: FAILED
* [Build Logs](https://fegcnxu6da.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=50582d0b-9246-4353-b8d0-f920fbebcc89%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

File: src/sagemaker/jumpstart/cache.py (hunk 3 above)
Comment (Captainia) [reply src/sagemaker/jumpstart/cache.py:149]: Thanks! good suggestion

**Comment by mufaddal-rohawala** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: sagemaker-python-sdk-notebook-tests
* Commit ID: ab8ee02a103cc8102540fabe8d3b5471e13f09e4
* Result: SUCCEEDED
* [Build Logs](https://bkqkixfb59.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=dd6f0b02-cdee-4245-b13f-44a0bff2f626%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by mufaddal-rohawala** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: sagemaker-python-sdk-slow-tests
* Commit ID: ab8ee02a103cc8102540fabe8d3b5471e13f09e4
* Result: SUCCEEDED
* [Build Logs](https://omfrydptb9.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=8ae4980d-d662-415c-8124-02a4cb42df5b%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by mufaddal-rohawala** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: sagemaker-python-sdk-pr
* Commit ID: ab8ee02a103cc8102540fabe8d3b5471e13f09e4
* Result: FAILED
* [Build Logs](https://abkt2mlh3k.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=ab840404-8e0b-454c-84d8-f258451a3629%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

File: src/sagemaker/jumpstart/utils.py [hunk 7]
```diff
@@ -752,16 +773,25 @@ def is_valid_model_id(
...
-
     models_manifest_list = accessors.JumpStartModelsAccessor._get_manifest(
-        region=region, s3_client=s3_client
+        region=region, s3_client=s3_client, model_type=enums.JumpStartModelType.OPENSOURCE
+    )
+    open_source_model_id_set = {
+        model.model_id for model in models_manifest_list
+    }
+
+    proprietary_manifest_list = (
```
Comment (akozd) [inline src/sagemaker/jumpstart/utils.py:783]: out of scope for this PR, but perhaps it is better to store this structure as a dict, and then if you ever need a list, call `.values()` on the dict.

**Pull request closed without merging**